rfcs/60-network-error.md

```diff
@@ -0,0 +1,204 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-05-11
+---
+
+# AOS-60: Network Error and Retryable Server-side Errors
+
+## Background
+
+We can classify all errors by the root cause of error:
+
+- User-side: incapability, wrong usage, pass an append object into multipart operations (like `failed to send a request to the storage system`)
+- Network-side: Network related error (like `failed to receive a response from the storage system`)
+  - explicit error: connection refuse, connection reset
+  - implicit error: timeout
+- Service-side: not exist, permission denied, already exists... (like `received an error response from the storage system`)
+
+All the errors will go into `formatError`, and be converted into either a defined error or an unexpected error. Currently, we have only defined some user-side errors and server-side errors. 
+
+For errors which are returned by SDK, but not belong to the SDK's error type, they are either belong to user-side or network-side. 
+
+```go
+func formatError(err error) error {
+	if _, ok := err.(services.AosError); ok {
+		// client-side
+		return err 
+	}
+
+	var e *qserror.QingStorError
+	if !errors.As(err, &e) {
+		// client-side or network-side
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+
+	// server-side
+	switch e.Code {
+	case "permission_denied":
+		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, e)
+	case "object_not_exists":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, e)
+	default:
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+}
+```
+
+Network errors are common and should be expected. Besides, there's a common need to retry a network error. So we should provide network error codes. 
+
+It will be better if a good retry mechanism can be built on top of network errors.
+
+## Proposal
+
+So I propose to add the following global error codes:
+
+Network-side: 
+
+- `ErrNetworkTemporary`: context.canceled, connection timeout, no response for some time 
+
+And use the following function to judge and convert.
+
+```go 
+func IsErrNetworkTemporary(err error) bool {
+	if errors.Is(err, io.ErrUnexpectedEOF) {
+		return true
+	}
+
+	if _, ok := err.(interface{ Temporary() bool }); ok && e.Temporary() {
+		return true
+	}
+
+	var urlErr *url.Error
+	if errors.As(err, &urlErr) {
+		retriable := []string{"connection refused", "connection reset"}
+		for _, s := range retriable {
+			if strings.Contains(e.Error(), s) {
+				return true
+			}
+		}
+	}
+
+	var netErr *net.OpError
+	if errors.As(err, &netErr) && netErr.Op == "dial" {
+		return true
+	}
+	
+	return false
+}
+```
+
+Retryable server-side errors:
+
+- `ErrServerInternal`: HTTP 5xx
```

Comment:
For IO-related operations, there is an extra error: the input stream could only be consumed once which could not be retryable even we got a retryable server-side error.

So maybe we can't provide/implement any retry logic, just tell users whether they can retry their operations?

Reply:
I didn't consider deeply how to implement a retry logic in this proposal, but only try to help identify retriable errors (and only gave one simple `ErrServerInternal`).

Reply:
Oh, I just seem to know what you mean. Do you mean that the user fails to upload a big object, then we cannot retry because the input stream is consumed? 

This is really a problem for operation-level retry, and we are also not able to provide request-level retry since we use SDKs. 🤔